scenarios/smart_trailer/digital_twin_providers/trailer_properties_provider/src/trailer_properties_provider_impl.rs

```diff
@@ -0,0 +1,256 @@
+pub struct TrailerPropertiesProviderImpl {
+    pub data_stream: watch::Receiver<i32>,
+    pub min_interval_ms: u64,
+    entity_map: Arc<RwLock<HashMap<String, Vec<TopicInfo>>>>,
+}
+
+/// Create the JSON for the trailer weight property.
+///
+/// # Arguments
+/// * `trailer_weight` - The trailer weight value.
+fn create_property_json(trailer_weight: i32) -> String {
+    let metadata = Metadata {
+        model: trailer_v1::trailer::trailer_weight::ID.to_string(),
+    };
+
+    let property: TrailerWeightProperty = TrailerWeightProperty {
+        trailer_weight,
+        metadata,
+    };
+
+    serde_json::to_string(&property).unwrap()
+}
+
+/// Publish a message to a MQTT broker located.
+///
+/// # Arguments
+/// `broker_uri` - The MQTT broker's URI.
+/// `topic` - The topic to publish to.
+/// `content` - The message to publish.
+fn publish_message(broker_uri: &str, topic: &str, content: &str) -> Result<(), String> {
+    let create_opts = mqtt::CreateOptionsBuilder::new()
+        .server_uri(broker_uri)
+        .client_id(MQTT_CLIENT_ID.to_string())
+        .finalize();
+
+    let client = mqtt::Client::new(create_opts)
+        .map_err(|err| format!("Failed to create the client due to '{err:?}'"))?;
+
+    let conn_opts = mqtt::ConnectOptionsBuilder::new()
+        .keep_alive_interval(Duration::from_secs(30))
+        .clean_session(true)
+        .finalize();
+
+    let _connect_response = client
+        .connect(conn_opts)
+        .map_err(|err| format!("Failed to connect due to '{err:?}"));
+
+    let msg = mqtt::Message::new(topic, content, mqtt::types::QOS_1);
+    if let Err(err) = client.publish(msg) {
+        return Err(format!("Failed to publish message due to '{err:?}"));
+    }
+
+    if let Err(err) = client.disconnect(None) {
+        warn!("Failed to disconnect from topic '{topic}' on broker {broker_uri} due to {err:?}");
+    }
+
+    Ok(())
+}
+
+impl TrailerPropertiesProviderImpl {
+    /// Initializes provider with entities relevant to itself.
+    ///
+    /// # Arguments
+    /// * `data_stream` - Receiver for data stream for entity.
+    /// * `min_interval_ms` - The frequency of the data coming over the data stream.
+    pub fn new(data_stream: watch::Receiver<i32>, min_interval_ms: u64) -> Self {
+        // Initialize entity map.
+        let mut entity_map = HashMap::new();
+
+        // Insert entry for entity id's associated with provider.
+        entity_map.insert(
+            trailer_v1::trailer::trailer_weight::ID.to_string(),
+            Vec::new(),
+        );
+
+        // Create new instance.
+        TrailerPropertiesProviderImpl {
+            data_stream,
+            min_interval_ms,
+            entity_map: Arc::new(RwLock::new(entity_map)),
+        }
+    }
+
+    /// Handles the 'PUBLISH' action from the callback.
+    ///
+    /// # Arguments
+    /// `payload` - Payload sent with the 'PUBLISH' action.
+    pub fn handle_publish_action(&self, payload: CallbackPayload) {
```

Comment:
because unwraps will cause panics, consider converting this to return `Result` and handling the errors by using `?` or returning them